src/Acuminator/Acuminator.Utilities/DiagnosticSuppression/SuppressionManager.cs

```diff
@@ -88,53 +83,82 @@ public static void Init(ISuppressionFileSystemService fileSystemService, IEnumer
+			project.ThrowOnNull(nameof(project));
+			CheckIfInstanceIsInitialized();
 
-			var (assembly, message) = GetSuppressionInfo(semanticModel, diagnostic, cancellation);
+			//First check if file already exists to dismiss threads withou acquiring the lock
+			if (Instance._fileByAssembly.TryGetValue(project.Name, out var existingSuppressionFile) && existingSuppressionFile != null)
+				return existingSuppressionFile;
 
-			if (assembly == null)
+			lock (Instance._fileSystemService)
 			{
-				return false;
+				//Second check inside the lock if file already exists 
+				Instance._fileByAssembly.TryGetValue(project.Name, out existingSuppressionFile);
+				return existingSuppressionFile ?? AddNewSuppressionFileImpl();
 			}
 
-			var file = _fileByAssembly.GetOrAdd(assembly, (SuppressionFile)null);
-
-			if (file == null)
+			//---------------------------------------------Local Function--------------------------------------------------
+			SuppressionFile AddNewSuppressionFileImpl()
 			{
-				return false;
-			}
+				string suppressionFileName = project.Name + SuppressionFile.SuppressionFileExtension;
+				string projectDir = Instance._fileSystemService.GetFileDirectory(project.FilePath);
+				string suppressionFilePath = Path.Combine(projectDir, suppressionFileName);
 
-			if (file.GenerateSuppressionBase)
-			{
-                if (diagnostic?.Descriptor.DefaultSeverity != DiagnosticSeverity.Info)
-                {
-                    file.AddMessage(message);
-                }
+				//Create new xml document and get its text
+				System.Xml.Linq.XDocument newXDocument = SuppressionFile.NewDocumentFromMessages(Enumerable.Empty<SuppressMessage>());
+				string docText = GetXDocumentStringWithDeclaration(newXDocument);
 
-				return true;
+				//Add file to project and hard drive
+				var roslynSuppressionFile = project.AddAdditionalDocument(suppressionFileName, docText, filePath: suppressionFilePath);
+
+				if (!project.Solution.Workspace.TryApplyChanges(roslynSuppressionFile.Project.Solution))
+					return null;
+
+				var (suppressionFile, assembly) = Instance.LoadFileAndTrackItsChanges(suppressionFilePath, generateSuppressionBase: false);
+				Instance._fileByAssembly[assembly] = suppressionFile;
+				return suppressionFile;
 			}
+		}
 
-			if (!file.ContainsMessage(message))
-			{
+		public static bool SuppressDiagnostic(SemanticModel semanticModel, string diagnosticID, TextSpan diagnosticSpan,
+											  DiagnosticSeverity defaultDiagnosticSeverity, CancellationToken cancellation = default)
+		{
+			CheckIfInstanceIsInitialized();
+
+			if (!IsSuppressableSeverity(defaultDiagnosticSeverity))
 				return false;
-			}
 
-			return true;
+			var (fileAssemblyName, suppressMessage) = SuppressMessage.GetSuppressionInfo(semanticModel, diagnosticID, 
+																						 diagnosticSpan, cancellation);
+			if (fileAssemblyName.IsNullOrWhiteSpace() || !suppressMessage.IsValid)
+				return false;
+
+			lock (Instance._fileSystemService)
```

Comment:
Why do we need this lock?

Reply:
@metallmonster because the VS throws exception about access to file from another thread